libs/utils/trace_analysis.py

```diff
@@ -912,196 +810,354 @@ def plotCPUFrequencyResidency(self, cpus=None):
-            active_time.plot.barh(ax=axis,
+                                 figsize=(16,yrange));
+            active_time.plot.barh(ax=axes,
                                   color='r',
                                   legend=False,
-                                  figsize=(16,8),
-                                  fontsize=16);
-            axis.set_title('CPU{}'.format(cpu))
-            axis.set_xlabel('Time [s]')
-            axis.set_xlim(0, self.x_max)
-
-        if axis:
-            handles, labels = axis.get_legend_handles_labels()
-            plt.figlegend(handles,
-                          labels,
-                          loc='lower right')
-            plt.subplots_adjust(hspace=0.5)
+                                  figsize=(16,yrange));
+
+            axes.set_title('CPU{}'.format(cpu))
+
+            axes.grid(True);
+            if (idx+1 < n_cpus):
+                axes.set_xticklabels([])
+            else:
+                axes.set_xlabel('Time [s]')
+            axis.append(axes)
+
+        # Set same x-axes limit for all plots
+        for ax in axis:
+            ax.set_xlim(0, 1.05 * xmax)
+
+        # Put title on top of the figure. As of now there is no clean way to
+        # make the title appear always in the same position in the figure
+        # because figure heights may vary between different platforms
+        # (different number of OPPs). Hence, we use annotation
+        legend_y = axis[0].get_ylim()[1]
+        axis[0].annotate('OPP Residency Time',
+                         xy=(0, legend_y),
+                         xytext=(-50, 45),
+                         textcoords='offset points',
+                         fontsize=18)
+        axis[0].annotate('GREEN: Total',
+                         xy=(0, legend_y),
+                         xytext=(-50, 25),
+                         textcoords='offset points',
+                         color='g',
+                         fontsize=14)
+        axis[0].annotate('RED: Active',
+                         xy=(0, legend_y),
+                         xytext=(50, 25),
+                         textcoords='offset points',
+                         color='r',
+                         fontsize=14)
+        figname = '{}/{}cpu_freq_residency.png'.format(self.plotsdir,
+                                                       self.prefix)
+        pl.savefig(figname, bbox_inches='tight')
+
+    def _getClusterActiveSignal(self, cluster):
+        """
+        Build a square wave representing the active (i.e. non-idle) cluster
+        time, i.e.:
+            cluster_active[t] == 1 if at least one CPU is reported to be
+                                   non-idle by CPUFreq at time t
+            cluster_active[t] == 0 otherwise
+
+        :param cluster: list of CPU IDs belonging to a cluster
+        :type cluster: list(int)
+        """
+        idle_df = self.trace.df('cpu_idle')
+
+        cpu_active = {}
+        for cpu in cluster:
```

Comment:
I've also tried a completely different approach for the body of this loop... working at table level...
The idea is to:
1. inject a fake initial event if required
2. add a new column to the original DF which represents the active state
Here is the code:

```python
        for cpu in cluster:
            idle_cpu_df = idle_df[idle_df.cpu_id == cpu]

            # Inject intial status if required
            if idle_cpu_df.index[0] != 0:
                # Get a copy of the first cpu_idle events
                idle_cpu_0_df = idle_cpu_df.iloc[:1,:].copy(deep=True)
                # Reset its index to be at the origin [0.0]
                idle_cpu_0_df.reset_index(inplace=True)
                idle_cpu_0_df.loc[:,'Time'] = idle_cpu_0_df['Time'] - idle_cpu_0_df['Time']
                idle_cpu_0_df.set_index('Time', inplace=True)
                # The initial state is the opposite of the first cpu_idle event
                idle_cpu_0_df.loc[:,'state'] = idle_cpu_0_df['state'].apply(
                    lambda state : 0 if state == NONIDLE_STATE else NONIDLE_STATE)
                # Append the original set of events
                idle_cpu_0_df = idle_cpu_0_df.append(idle_cpu_df)
                idle_cpu_df = idle_cpu_0_df

            # Create a new CPU active signal
            idle_cpu_df.loc[:,'active'] = idle_cpu_df['state'].apply(
                lambda state : 1 if state == NONIDLE_STATE else 0)

            # Keep track of active states for this CPU
            cpu_active[cpu] = idle_cpu_df['active']
```

I would like to have something similar in the Trace module, as a Sanitize function for the cpu_idle DF.
Having such a column in the DF will allows to have the data on hand for many other plots...

However, for the time being it does not give much more advantages in performances (when measured via %timeti) and maybe the code complexity is not improved sensible.

Thus, for the time being, if we are able to simplify a bit the following if-else nesting... I would be happy to for with your version.